File: hashing/counting/python3/subarray_sum.py
from typing import Sequence
from collections import defaultdict
import numpy as np

def subarray_sum(nums: Sequence[int], k: int) -> int:
    """
    Given an integer array `nums` and an integer `k`, find the number of subarrays whose sum is equal to `k`.

    Parameters
    ----------
    nums : Sequence[int]
        A sequence of integers
    k : int
        An integer contraint

    Returns
    -------
    int
        Number of subarrays whose sum equal `k`
    """
    counts = defaultdict(int)
    # Empty subarray with sum of 0 has a count of 1
    counts[0] = 1
    subarray_with_sum_k_count = curr_sum = 0
    
    # O(n)
    for num in nums:
        # Add the next element
        curr_sum += num
        # Update the answer, which is the count of the prefix sum (curr_sum - k)
        subarray_with_sum_k_count += counts[curr_sum - k]
        # Keep track of the count of each prefix sum
        counts[curr_sum] += 1
        
    return subarray_with_sum_k_count

def main() -> int:
    
    for _ in range(5):
        k = np.random.randint(low=0, high=10)
        nums_size = np.random.randint(low=1, high=20)
        nums = np.random.randint(low=0, high=10, size=(nums_size, )).tolist()
        ans = subarray_sum(nums, k)
        print(f"Give the input {nums}, the number of subarrays whose sum is {k} is {ans}")
    
    return 0

if __name__ == "__main__":
    
    main()

File: hashing/counting/c++/subarray_sum/subarray_sum.cpp
// subarray_sum.cpp
#include "subarray_sum.hpp"
#include <vector>
#include <unordered_map>

int subarraySum(const std::vector<int> &nums, int k)
{
    // Initialize the hash map with the prefix sum of the empty subarray
    std::unordered_map<int, int> counts;
    counts[0]++;
    int currSum = 0, subarrayWithSumKCount = 0;

    for (const auto &num : nums)
    {
        // Add to the current prefix sum
        currSum += num;
        // Update the answer, which is the count of the prefix sum (curr_sum - k)
        subarrayWithSumKCount += counts[currSum - k];
        // Keep track of the count of each prefix sum
        counts[currSum]++;
    }

    return subarrayWithSumKCount;
}
